### ISSUE TITLE
simple change to avoid boxing in some callables

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-compiler#2323]
This is not a proposal to totally solve the problem of boxing with callables, but it's a fairly minimal change that would get us rather a decent way toward optimizing certain operations like `Array.each()`, `String.split()`, `String.trim()`, etc.

Currently, given this code:

```ceylon
(Integer i) => i+1
```

I get Java like this:

```java
new .com.redhat.ceylon.compiler.java.language.AbstractCallable<.ceylon.language.Integer>(.ceylon.language.Integer.$TypeDescriptor$, .com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor.tuple(false, false, -1, .ceylon.language.Integer.$TypeDescriptor$), "Integer(Integer)", (short)-1)
        @.com.redhat.ceylon.compiler.java.metadata.LocalDeclaration
        @.com.redhat.ceylon.compiler.java.metadata.Name("anonymous#0")
        @.com.redhat.ceylon.compiler.java.metadata.Method
        {
            
            @.java.lang.Override
            @.com.redhat.ceylon.compiler.java.metadata.Ignore
            public .ceylon.language.Integer $call$(final .java.lang.Object $param$0) {
                final long i = ((.ceylon.language.Integer)$param$0).longValue();
                return .ceylon.language.Integer.instance(i + 1L);
            }
        }
```

If, instead, we generated code like this:

```
new .com.redhat.ceylon.compiler.java.language.AbstractCallable<.ceylon.language.Integer>(.ceylon.language.Integer.$TypeDescriptor$, .com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor.tuple(false, false, -1, .ceylon.language.Integer.$TypeDescriptor$), "Integer(Integer)", (short)-1)
        @.com.redhat.ceylon.compiler.java.metadata.LocalDeclaration
        @.com.redhat.ceylon.compiler.java.metadata.Name("anonymous#0")
        @.com.redhat.ceylon.compiler.java.metadata.Method
        {
            
            @.java.lang.Override
            @.com.redhat.ceylon.compiler.java.metadata.Ignore
            public .ceylon.language.Integer $call$(final .java.lang.Object $param$0) {
                return .ceylon.language.Integer.instance(
                       $call$long$long(((.ceylon.language.Integer)$param$0).longValue()));
            }
            @.java.lang.Override
            @.com.redhat.ceylon.compiler.java.metadata.Ignore
            public long $call$long$long(final long $param$0) {
                final long i = $param$0;
                return i + 1L);
            }
        }
```

Then the native Java code in `Array` and `String` could call the primitive version of the method, thus completely avoiding boxing.

We only really care about the cartesian product of `{ Long, Float, Byte, Boolean, Character }` with itself, resulting in 25 methods on `Callable` since all the interesting operations have just one parameter. We _don't_ need to worry about optimizing callables with >1 parameter!

As a second step, we could also some optimize invocations from Ceylon code, which again doesn't look especially challenging. Instead of:

```java
fun.$call$((.java.lang.Object).ceylon.language.Integer.instance(1L)).longValue();
```

We would produce:

```java
fun.$call$long$long(1L);
```

This really doesn't look like a huge amount of work to me!

**Comment by quintesse** [comment]
I'd much rather make this part of the more generic solution we discussed some time ago where *any* class / interface with type parameters gets methods like these added.

The only thing to figure out would be how to decide *which* type parameters get this treatment.
A simple heuristic would be that for any class we add these methods for the first type parameter whose bounds include any of the basic types.
So by default we'd only do this for a single type parameter to avoid combinatorial explosion. But we'd allow people to override this with an annotation on the type parameters so they can decide to which and to how many of them (max 2? 3?) they want it applied.

**Comment by gavinking** [comment]
> I'd much rather make this part of the more generic solution we discussed some time ago where any class / interface with type parameters gets methods like these added.

Well that's a _much, much_ harder problem, and one that we're unlikely to work on in the immediate future.

This is something we can bite off, that will make a huge difference to the performance of certain common problems. It's low-hanging fruit.

**Comment by FroMage** [comment]
Actually if we start using method handles we can work unboxed entirely, so this trick may just not be necessary at all.

**Comment by tombentley** [comment]
We've no idea what the overall perf impact of using method handles would be. Obviously we can seek to use `invokeExact()` everywhere to avoid all that boxing, but the actual invocation might not work out to be as fast as the `invokevirtual` we have currently.

**Comment by FroMage** [comment]
Well, no, but given that it saves us lots of inner classes, and possibly lots of instances too, with constant lambdas, and is the path taken by Java 8, I can only assume it'd perform a hell of a lot better than what we have right now.

**Comment by tombentley** [comment]
You're not comparing apples with apples. Our `Callable` is not the same as a single `MethodHandle`. You could probably make something like a `Callable` using an array of `MethodHandle`s (one for each arity of `Callable`), with machinery to select the right handle and `invokeExact()` it.

**Comment by FroMage** [comment]
Well, I'm comparing what we can do with method handles when we don't have any variable arity. You're right that the variable arity part is still an unknown.

**Comment by quintesse** [comment]
> Well that's a much, much harder problem

I don't think it is actually, at least not hard as in "serialization" hard or "meta-model" hard. It's just a bit of work, that's all. Because the hardest part is not generating those methods, the hardest part is finding all the parts in the code which we would need to change when calling these new methods. And that part would be similar in complexity in both cases.

So given that this is for 1.3 anyway why not do it right? :)
(And I'm volunteering coz I've wanted to do this for ages already)
